Version 1.2.0

originalfoo opened the pull request:
Major update:

Fixes #18  - log should show num occurrences of assembly
Fixes #14 - struct for mod + assembly info
Fixes #13 - refactoring folders + classes
Fixes #12 - log MD5 of assemblies
Fixes #11 - indicate if assembly is loaded, indicate if mod is enabled
Fixes #10 - show detailed results in mod options screen
Fixes #9 - scan mod folder if it's `PluginInfo` doesn't list the assembly

originalfoo committed:
Version 1.2.0

Major update

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Settings.cs:
```diff
@@ -0,0 +1,124 @@
+    /// </summary>
+    public class Settings {
+        // Markers used in checkbox labels
+        private const string MARKER_BLANK = " ";
```
The .NET naming convention suggests using `PascalCase` for constants.

originalfoo replied:
Will fix as part of #19 for v1.2.1

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Settings.cs:
```diff
@@ -0,0 +1,124 @@
+        /// </summary>
+        /// 
+        /// <param name="helper">UI helper for building settings screen.</param>
+        public static void CreateUI(UIHelperBase helper) {
```
The .NET style guide suggests using the BSD (Allman) style of the braces: the opening brace must be placed on an empty line.

originalfoo replied:
Will deal with this as part of #19 at later date.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Settings.cs:
```diff
@@ -0,0 +1,124 @@
+            UIHelperBase group;
+
+            if (CacheScan(
+                out Dictionary<string, int> duplicates,
```
Consider `out var`.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Settings.cs:
```diff
@@ -0,0 +1,124 @@
+                Disable((UICheckBox)group.AddCheckbox("= Assembly is loaded in to app domain", true, NoOp));
+                Disable((UICheckBox)group.AddCheckbox("= Assembly not loaded", false, NoOp));
+
+                foreach (KeyValuePair<string, int> duplicate in duplicates) {
```
Consider `foreach(var duplicate in duplicates)`

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Settings.cs:
```diff
@@ -0,0 +1,124 @@
+
+                    if (duplicate.Value > 1) {
+
+                        if (occurrences.TryGetValue(duplicate.Key, out List<ModAssembly> mods)) {
```
Consider `out var`

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Settings.cs:
```diff
@@ -0,0 +1,124 @@
+        /// </summary>
+        /// 
+        /// <param name="_">Ignored paramter.</param>
+        internal static void NoOp(bool _) {
```
The parameter names should not be presented as discards (underscore).
Instead of a dummy method, just provide a lambda that does nothing: dummy empty methods in a class are considered as design flaw.

originalfoo replied:
Removed `NoOp` method and replaced event handlers with with `default` which I believe creates empty lambda based on delegate definition.

originalfoo replied:
Ok, didn't work in-game, replaced with empty lambda.

dymanoid replied:
`default` will always generate a `null` for reference types. Lambdas are reference types.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Settings.cs:
```diff
@@ -0,0 +1,124 @@
+        private const string MARKER_ENABLED = "*";
+
+        // Cache of scan results
+        private static Dictionary<string, int> duplicates_;
```
The .NET naming convention suggests either `PascalCase` or `camelCase` for static fields. The underscore should be used as a prefix for `_instanceFields`.

originalfoo replied:
Will fix as part of #19 for v1.2.1

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Settings.cs:
```diff
@@ -0,0 +1,124 @@
+            UIHelperBase helper,
+            string assemblyName,
+            int loaded,
+            List<ModAssembly> list) {
```
`list` could be `IReadOnlyCollection<ModAssembly>`

originalfoo replied:
Updated but curious to know what the benefit is?

originalfoo replied:
Ah, didn't work - we're on .Net 3.5 and the package required for `IReadOnlyCollection` requires later version. :(

dymanoid replied:
Sorry, I always forget we're stuck in the Stone Age with Unity 5...

krzychu124 commented:
I'll test it later today 😉 I need 😴 💤

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Settings.cs:
```diff
@@ -0,0 +1,124 @@
+            
+            if (duplicates_ == null) {
+                duplicatesFound_ = Assemblies.Scan(
+                    out Dictionary<string, int> dup,
```
Consider `out var`

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Structs/ModAssembly.cs:
```diff
@@ -0,0 +1,50 @@
+namespace DuplicateAssemblyScanner.Structs {
+    using DuplicateAssemblyScanner.Util;
+    using System.Reflection;
+    using static ColossalFramework.Plugins.PluginManager;
```
Prefer not to use `using static`. This might confuse the readers. Only use `using static` for `static` classes containing `public const` fields, or better do not use `using static` at all.

originalfoo replied:
Doing so breaks the `<see cref="PluginInfo.modPath"/>` references in comments.

originalfoo replied:
Removed the `see` comments.

dymanoid replied:
You just can use `<see cref="ColossalFramework.Plugins.PluginManager.PluginInfo.modPath"/>`

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Structs/ModAssembly.cs:
```diff
@@ -0,0 +1,50 @@
+        /// <summary>
+        /// Path of containing folder for the mod (<see cref="PluginInfo.modPath"/>).
+        /// </summary>
+        public string ModPath;
```
Do not introduce `public` non-const fields. Use properties instead.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Structs/ModAssembly.cs:
```diff
@@ -0,0 +1,50 @@
+    /// <summary>
+    /// Contains details of an assembly within a mod.
+    /// </summary>
+    public struct ModAssembly {
```
Do not introduce `struct`s without a *real* need for that. Here, use a `class` instead.
Mutable structs are **evil**. Only use mutable structs for high-perf reasons when you need a continuous memory storage for a large array of data (obviously, this is not the case here).

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Util/Md5.cs:
```diff
@@ -0,0 +1,56 @@
+        /// <param name="removeHyphens">If <c>true</c>, hyphens will be removed form the returned string.</param>
+        /// 
+        /// <returns>Returns <c>true</c> if successful, otherwise <c>false</c>.</returns>
+        internal static bool TryGetHash(string fullPath, out string md5Hash, bool removeHyphens = false) {
```
`out` parameters should be placed at the end of the method parameter list; in this case, the `removeHyphens` parameters will become a required one. Create an overload without `removeHyphens` if you wish.

originalfoo replied:
The param was no longer needed so removed it.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Util/MemberValue.cs:
```diff
@@ -0,0 +1,95 @@
+        public static T GetValue<T>(object instance, MemberInfo member) {
+            return member.MemberType switch
+            {
+                MemberTypes.Field => (T)(member as FieldInfo)?.GetValue(instance),
```
The null check is not necessary since `MemberType.Field` implies that `member` is `FieldInfo`.

originalfoo replied:
Is this also the case for the `Method` and `Property`?

dymanoid replied:
I'd completely rewrite the `switch` expression by using the pattern matching:
```C#
return member switch
{
    FieldInfo field => (T)field.GetValue(instance),
    MethodInfo method => (T)method.Invoke(instance, null),
    PropertyInfo property => (T)property.GetValue(instance, null),
    _ => default,
};
```
Note, however, that this method is unsafe. It will throw if `T` happens to be a `struct` and `member` will be a `void`-returning method.

originalfoo replied:
Class is no longer used; removed.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Util/MemberValue.cs:
```diff
@@ -0,0 +1,95 @@
+            object instance,
+            string name,
+            out T value,
+            BindingFlags flags = DefaultBindingFlags) {
```
Do not use default parameters for `public` methods.

originalfoo replied:
Made method `private`.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Util/Mods.cs:
```diff
@@ -0,0 +1,197 @@
+        /// <returns>Returns <c>true</c> if successful, otherwise <c>false</c>.</returns>
+        private static bool TryGetAssemblyPath(PluginInfo mod, string assemblyName, out string fullPath) {
+            fullPath = Path.Combine(mod.modPath, $"{assemblyName}.dll");
+            try {
```
No need for `try-catch` for `File.Exists`: it does not throw.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/Util/Mods.cs:
```diff
@@ -0,0 +1,197 @@
+        private static bool TryGetAssemblyPath(PluginInfo mod, string assemblyName, out string fullPath) {
+            fullPath = Path.Combine(mod.modPath, $"{assemblyName}.dll");
+            try {
+                if (File.Exists(fullPath)) {
```
`return File.Exists(fullPath);`

dymanoid commented on DuplicateAssemblyScanner/VersionInfo.cs:
```diff
@@ -9,11 +9,11 @@
 
 // Ownership information
-[assembly: AssemblyCompany("Cities Skylines Mods")]
+[assembly: AssemblyCompany("github.com/CitiesSkylinesMods")]
```
This is not nice. The company name shouldn't be an URL.

originalfoo committed:
Refactored logging - fixes #18

originalfoo committed:
Changes based on dymanoid review

originalfoo committed:
Logging OnSettingsUI again

Because this https://github.com/CitiesSkylinesMods/DuplicateAssemblyScanner/issues/8#issuecomment-595654252

originalfoo committed:
Simplified `using` in Log.cs

originalfoo committed:
Update solution items

originalfoo committed:
Minor tweak to log file formatting

originalfoo committed:
Dimmed non-loaded assembly checkboxes

dymanoid commented:
As a general note regarding your git commits. Consider reading [this great blog post](https://chris.beams.io/posts/git-commit/) about the git commit messages. I'd even recommend to publish a link to this in your contribution guide. The blog post explains how to make your git history readable and clean.

- Limit the subject line to 50 characters
- Use the imperative mood in the subject line (`Simplify`, not `Simplified` / `Simplifying` / `Simplification`)
- Capitalize the subject line
- Do not end the subject line with a period
- Use the body to explain *what* and **why** vs. *how* (*if there is something the developers need to know besides the subject line*)
- Separate subject from body with a blank line (*if there is a body*)
- Wrap the body at 72 characters (*if there is a body*)

originalfoo committed:
Improve the build environment

In `Release` builds, constant TRACE is not wanted. Replace it with RELEASE.

References do not specify assembly version, public key, etc. Add that detail.

The post-build script contains hard-coded paths and is somewhat opaque as to
how the local C:SL folders are structured. Make use of macros and environment
variables to generalise the script, and echo folder paths to build console as
visual confirmation that the correct paths are being used.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/DuplicateAssemblyScanner.csproj:
```diff
@@ -42,33 +42,51 @@
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Assembly-CSharp">
+    <Reference Include="Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null">
```
Why would you think this is important here?

originalfoo replied:
Following the approach taken in https://github.com/CitiesSkylinesMods/PathfindSandbox/blob/master/PathfindSandbox/PathfindSandbox.csproj

dymanoid replied:
You only need to request a specific version if there is a requirement to do so. (E.g. your code depends on a feature that is only available in that specific version, or you know a major bug in other versions so you want to avoid it). In all other cases, just use the assembly name, without a version.
Specifying assembly versions and key tokens for the system references (like `mscorlib`) is meaningless, because there is only *one* version of that assembly in a given runtime anyway (including Unity/Mono). You have no option to get a different one.
Requesting version `0.0.0.0` is quite awkward too (that is because Unity generates that dummy version stub on building the game assembly).

originalfoo replied:
Will sort this out in separate project, I'm splitting out a load of common props from the csproj file so they can be shared across multiple projects/repos.

dymanoid commented on DuplicateAssemblyScanner/DuplicateAssemblyScanner/DuplicateAssemblyScanner.csproj:
```diff
@@ -42,33 +42,51 @@
     </Reference>
-    <Reference Include="ColossalManaged">
+    <Reference Include="ColossalManaged, Version=0.3.0.0, Culture=neutral, PublicKeyToken=null">
       <HintPath>$(MangedDLLPath)\ColossalManaged.dll</HintPath>
```
Maybe `Managed`, not `Manged`?

originalfoo committed:
Fix typo 'Manged' in csproj

`MangedDLLPath` should be `ManagedDllPath`.

originalfoo committed:
Move `ModAssembly.cs` out of `Structs` folder

ModAssembly is a class, not a struct.

The Structs folder is no longer required.

originalfoo committed:
Delete redundant MemberValue.cs

originalfoo committed:
Remove MemberValue from csproj
